compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java

```diff
@@ -180,6 +231,45 @@ private void addPlatformLibraryPaths(PackageContext packageContext,
         }
     }
 
+    private static boolean hasEmptyIdOrVersion(JarLibrary entry) {
+        return entry.groupId().isEmpty() || entry.artifactId().isEmpty() || entry.version().isEmpty();
+    }
+
+    private boolean isUsedDependency(JarLibrary otherJarDependency, Set<String> usedNativeClassPaths,
+                                     int totalJarDependencies) {
+        String pkgName = otherJarDependency.packageName().get();
+        if (totalJarDependencies != 1) {
+            return true;
+        }
+        if (isWhiteListedPkg(pkgName)) {
+            return true;
+        }
+        if (usedNativeClassPaths == null) {
+            return false;
+        }
+        try (JarFile jarFile = new JarFile(otherJarDependency.path().toFile())) {
+            for (String classPath : usedNativeClassPaths) {
+                ZipEntry usedClassEntry = jarFile.getJarEntry(classPath);
+                if (usedClassEntry != null) {
+                    return true;
+                }
+            }
+            return false;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean isWhiteListedPkg(String pkgName) {
+        return pkgName.equals("ballerina/observe") || pkgName.equals("ballerinai/observe") || isDriverPkg(pkgName);
+    }
+
+    // Driver pkgs are pkgs such as "ballerinax/mysql.driver".
+    // These pkgs contain only native jars without any source code.
+    private boolean isDriverPkg(String pkgName) {
+        return pkgName.startsWith("ballerinax/") && pkgName.endsWith(".driver");
```

Comment:
Use constant here and above packages.

Reply:
Fixed in 7ac189d8d68a32b7f7175045b7c8a1db5d78f3ef.